File: ProblemSet/464 · 整数排序 II/java/464 · 整数排序 II 快速排序 双向扫描 法1.java
import java.util.Random;

public class Solution {
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        quickSort(A, 0, A.length - 1);
    }
    
    private void quickSort(int[] nums, int left, int right) {
        if (left < right) {
            int pos = partition(nums, left, right);
            quickSort(nums, left, pos - 1);
            quickSort(nums, pos + 1, right);
        }
    }

    private int partition(int[] nums, int left, int right) {
        int pivotIndex = new Random().nextInt(right - left + 1) + left;
        swap(nums, left, pivotIndex);
        int low = left + 1, high = right, pivot = nums[left];
        while (true) {
            while (low <= high && nums[low] < pivot) ++low;
            while (low <= high && nums[high] > pivot) --high;
            if (low < high) {
                swap(nums, low, high);
                ++low;
                --high;
            }
            else break;
        }
        swap(nums, left, high);
        return high;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

}

File: ProblemSet/464 · 整数排序 II/python/464 · 整数排序 II 快速排序 双向扫描 法1.py
import random

class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        self.n = len(A)
        self.quickSort(A, 0, len(A) - 1)

    def quickSort(self, nums, left, right):
        if left < right:
            pos = self.partition(nums, left, right)
            self.quickSort(nums, left, pos - 1)
            self.quickSort(nums, pos + 1, right)

    
    def partition(self, nums, left, right):
        pivotIdx = random.randint(0, right - left) + left
        nums[left], nums[pivotIdx] = nums[pivotIdx], nums[left]
        low, high, pivot = left + 1, right, nums[left]
        while True:
            while low <= high and nums[low] < pivot:
                low += 1
            while low <= high and nums[high] > pivot:
                high -= 1
            if low < high:
                nums[low], nums[high] = nums[high], nums[low]
                low += 1
                high -= 1
            else:
                break
        nums[left], nums[high] = nums[high], nums[left]
        return high